File: arrays_and_hashing/0347_top_k_frequent_elements.cpp
// https://leetcode.com/problems/top-k-frequent-elements/

#include <vector>
#include <unordered_map>

class Solution {
public:
    std::vector<int> topKFrequent(std::vector<int>& nums, int k) {
        std::unordered_map<int, int> freq;

        for (int& num: nums) {
            freq[num]++;
        }

        std::vector<std::vector<int>> buckets(nums.size() + 1);

        for (const auto& [num, count]: freq) {
            buckets[count].push_back(num);
        }

        std::vector<int> result;

        for (int i = buckets.size() - 1; i >= 0; --i) {
            for (int num: buckets[i]) {
                result.push_back(num);

                if (result.size() == k) {
                    return result;
                }
            }
        }
        return result;
    }
};

File: arrays_and_hashing/0347_top_k_frequent_elements.py
# https://leetcode.com/problems/top-k-frequent-elements/

from collections import Counter


class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        counter = Counter(nums).most_common(k)
        result = [num for num, _ in counter]
        return result
